### PR TITLE
[test-generation] Allow generation of generic structs and functions

### PR DESCRIPTION
**Pull request by tzakian**
This PR adds generics to test generation. 

This has a number of effects, and I'll try to explain some of the high-level ideas here. I'll first explain the reasoning for structs, and the reasoning for function calls is similar (although in many aspects, easier). 

#### Current Status:
Ready for review.

## Structs/Resources
We shall assume throughout that we have the following struct definitions:
```rust
struct A<T: All> {
    a: T,
} 

struct C<T: Unrestricted, L: Resource> {
    c: L
}

struct D<T: Unrestricted, L: Unrestricted> {
    d: T
    e: B
}

struct E<T: Unrestricted> {
   f: T,
   g: T,
}
```

Let's first take a look at some examples:
#### 1: Inferring Instantiations (complete case)
We have a (abstract) stack that looks like the following: `[..., U64]`. Then we need to _infer_ the instantiation of `A` to be `U64`;
```rust
Pack(A, <u64>)
```

Similarly, if we have a abstract stack of `[..., T]`, then we need to _infer_ the instantiation:
```rust
Pack(A, <T>)
```

As well, if we have an abstract stack like the following: `[..., T, A<T>]` we need to _infer_ the instantiation
```rust
Pack(D, <T, A<T>>) 
```
This work is done by the `Subst` struct and the method `check_and_add` in `transitions.rs`, and in the `stack_struct_satisfies_signature`. There is also sub-kinding that is handled -- making sure that the kinds of `T` and `A<T>` are `subkinds` of the expected kinds for the parameters.

We also make sure in this substitution creation that the substitution works across fields of the same type e.g. if we had the following stack `[..., Bool, U64]` we could not infer an instantiation for `E`:
```rust
check_and_add(subst = {}, Bool, T) ~~> subst = {T |-> Bool};
check_and_add(subst = {T |-> Bool}, U64, T) ~~> U64 == subst(T) ~~> U64 == Bool ~~> false
```

#### 2: Inferred, but incomplete instantiations

Take the following abstract stack: `[..., L:R]`. Then according to the rules specified above, we can generate a `Pack` of `C`. However, we have a problem: the types on the abstract stack are not enough for us to fully determine the full instantiation of `C` -- we have a partial instantiation, with `T` being free. In this case we simply choose a type to instantiate with (currently `U64`, but we could do something different although I'm not sure what benefit we would get from it). We don't handle free resource instantiations...yet. 

#### 3: Checked

The control flow graph that is generated for bytecode generation specifies starting states for the start and end of every basic block. One of these properties are the sets of locals that are `Available` or `Unavailable` (have had their value moved out of them). In such a setting, we may need to generate a bytecode sequence that inhabits a _particular_ type. In most cases "checked" vs "inferred" cases matchup. However, in the case of a "free" type variable we run into issues--in the inferred context, the typing information is flowed from the abstract state to the bytecode generation, whereas in this case (the inhabitation of a _particular_ type) we need to flow the information the other way around -- from the bytecode generation into the abstract state and instantiation. If we only flowed from the abstract state, we could not generate a local of type `C<Bool, L>` (where `L` has kind `Resource`)--we could succeed in choosing `L` from the abstract stack, but when it came time to choose the instantiation for the "`T`" parameter of `C` we simply don't have enough information in inferred mode. This is where the `exact` flag comes into play for the summaries -- it serves as a flag to determine the direction that typing information should flow. 

#### 4: Unpack
This is relatively straightforward -- we grab the instantiation of the struct to be unpacked from the abstract stack, and perform type substitution on each field. We likewise use this as the instantiation for the `Unpack` instruction. 

#### General 

In all of the cases above, the effects to the stack are dependent upon the instantiation chosen. Therefore the effects for the instructions summaries have been updated to encapsulate this e.g. `Call` and `Pack` have effects that are dependent upon the instantiation chosen (since e.g. in the `Pack` case the instantiation could cause a non-resource or resource to be pushed on the stack). 

## Functions
These are relatively straightforward, and follow the same kind of path as structs.

# The Details

#### Addition of `InstantiableModule`
Since we can now create instantiations in the process of generating bytecode, we need to be able to intern these, and have the ability to (1) see if the instantiation (a `Vec<SignatureToken>`) has already been interned, and if so the index of it in the table, (2) intern an instantiation, and return the index of that instantiation (in the `LocalsSignaturePool`), and (3) get the instantiation for a `LocalsSignatureIndex`. We do this by wrapping the module with a hashmap of instantiation to `LocalsSignatureIndex` and likewise have a reverse lookup table of `LocalsSignatureIndex` to instantiation (represented as a vector). Since the abstract state is thrown away from one block to another, we need to `instantiate` the module at the end of every block, and use this new module for the next block.

#### Addition of `instantiation` field
This represents the kinds for the type parameters of the function that we are in. e.g. if we are generating bytecode within a function `f<T: Unrestricted, L: All>(...) { ... }` the `instantiation` field will be `[Unrestricted, All]`.

#### Addition of `control_flow_allowed`
We do _not_ want to automatically generate control instructions. However, we also need to be able to manually add these, and apply their effects to the abstract state, so we have this flag that we manually set to allow us to insert these operations, but this won't be set when we are trying to generate random instructions and therefore they will never be selected as possible instructions.

## Changes to transitions

The main change is that we now need to determine the instantiations for instructions (or, in the checked mode for locals, verify that the state is amenable to the particular instruction/instantiation combination). To infer the instantiation we have the `check_and_add` function in `Subst`. This allows us to build up an instantiation mapping of `<ty param offset> -> type on stack`. We then use this as the final mapping for creating an instantiation (after filling in "free" type parameters). We also use this to verify the substitution works for all fields. 

We also have this idea of `exact`ness -- this is specifying which way the information flow is going "inexact` is saying that type info is inferred from the stack, and "exact" is saying that it is being provided, or pushed down from generation.

## Key things to keep in mind
* We cannot narrow a stack type (either by type, or kind). But, we can narrow an instantiation kind, or type to match the type on the stack.
* We need to flow type information both from the abstract state "what instantiation should I have for this instruction" and we also need to flow the type information the other way "I need an instruction with exactly this type, make sure it works and apply it to the state".

### Things we don't do
* Inhabitation of locals with a generic parameter...This is "future work"(!?!?). Tell me how to generate a value of a generic type `T` and I'll tell you you're wrong (or, you're right, and I'll be happy to help you apply for a Turing award....).
* Any types of references -- I turned these instructions off until we get the borrow graph in. Without we can't really have references in the presence of structs (etc).

## Future work
* Need to optimize this, there's a lot of low-hanging fruit here for cleanup and optimization. But that's for a future PR.
* We can sometimes generate functions bodies that are too large -- we sometimes cause a panic in the `control_flow_graph` in the bytecode verifier (and also during serialization of the blocks internally sometimes as well) since we overflow a `u16`. We need to have checks for this and fail elegantly in these cases (especially in the bytecode verifier!)
* Better bounding of the value and call stacks -- we sometimes overflow these. We need to encode back pressure more effectively off of these during the generation phase.

### CODE REVIEW

File: language/tools/test-generation/src/abstract_state.rs [hunk 1]
```diff
@@ -85,6 +87,102 @@ impl AbstractValue {
...
+            None => {
+                panic!("Unable to get instantiation at offset: {:#?}", index);
+            }
+        }
+    }
+
+    /// Consumes self, and adds the instantiations that have been built up to the underlying
+    /// module, and returns the resultant compiled module.
+    pub fn instantiate(self) -> CompiledModuleMut {
+        let index = self.module.locals_signatures().len();
```
Comment (dariorussi) [inline language/tools/test-generation/src/abstract_state.rs]: just out of curiosity and to see what I am missing... why not use `instance_for_offset` directly here?

File: language/tools/test-generation/src/transitions.rs [hunk 2]
```diff
@@ -9,24 +9,110 @@ use crate::{
...
+    pub fn new() -> Self {
+        Self {
+            subst: HashMap::new(),
         }
-        None => index < state.stack_len(),
     }
+
+    /// NB that the position of arguments here matters. We can build a substitution if the `instr_sig`
+    /// is a type parameter, and the `stack_sig` is a concrete type. But, if the stack signature is a
+    /// concrete type, but the stack signature is a type parameter, they cannot unify and no
```
Comment (dariorussi) [inline language/tools/test-generation/src/transitions.rs]: `if the stack signature is a` -> `if the instruction signature is a`?

File: language/tools/test-generation/src/transitions.rs [hunk 3]
```diff
@@ -43,6 +129,43 @@ pub fn stack_kind_is(state: &AbstractState, index: usize, kind: Kind) -> bool {
     false
 }
 
+/// Determine if the abstract value at `index` has a kind that is a subkind of the kind for the
```
Comment (dariorussi) [inline language/tools/test-generation/src/transitions.rs]: the usage of **kind** here is **kind** of funny (pun intended) ;)

File: language/tools/test-generation/tests/local_instructions.rs [hunk 4]
```diff
@@ -161,26 +161,27 @@ fn bytecode_immborrowloc_no_local() {
...
-fn bytecode_immborrowloc_local_unavailable() {
-    let mut state1 = AbstractState::new();
-    state1.local_insert(
-        0,
-        AbstractValue::new_primitive(SignatureToken::U64),
-        BorrowState::Unavailable,
-    );
-    common::run_instruction(Bytecode::ImmBorrowLoc(0), state1);
-}
+// TODO: Turn back on when borro graph and references are allowed
```
Comment (dariorussi) [inline language/tools/test-generation/tests/local_instructions.rs]: typo `borro` -> `borrow`

File: language/tools/test-generation/src/abstract_state.rs (hunk 1 above)
Comment (tzakian) [reply language/tools/test-generation/src/abstract_state.rs]: This is actually a really good spot, and I think I'm being stupid here -- I think we can just use `instance_for_offset`. I'll take a look and see, but I think it should be exactly what is needed.

File: language/bytecode-verifier/src/control_flow_graph.rs [hunk 5]
```diff
@@ -102,6 +102,7 @@ impl VMControlFlowGraph {
     }
 
     fn is_end_of_block(pc: CodeOffset, code: &[Bytecode], block_ids: &Set<BlockId>) -> bool {
+        // XXX: Overflow of PC is possible here!
```
Comment (shazqadeer) [inline language/bytecode-verifier/src/control_flow_graph.rs]: Did we resolve this issue based on the discussion before the holidays?  The tentative conclusion was that the bytecode generator is not modeling the serializer.

Comment (tzakian) [reply language/bytecode-verifier/src/control_flow_graph.rs]: We did indeed resolve this, I just haven't gotten around to update yet, but I'll be removing this comment. 

One thing worth thinking about, but that we don't need to address now, is whether or not the bytecode verifier should assume these type of preconditions (e.g. that the code passed to it it also passed the deserializer).

Comment (shazqadeer) [reply language/bytecode-verifier/src/control_flow_graph.rs]: Good point.  The "bytecode verifier" is organized as a sequence of passes.  These passes are partially-ordered in that a pass depends on all passes to successfully complete before being invoked.  This organization is likely inevitable.  The bytecode deserializer could be either thought of as the first pass or we could implement a "pass 0" in the bytecode verifier that does the checks performed by the deserializer.  

@dariorussi : What is your opinion?

**Pull request closed without merging**